src/client/common/installer/backupPipInstaller.ts

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { inject, injectable, named } from 'inversify';
+import { CancellationToken, Disposable, OutputChannel, ProgressLocation, ProgressOptions } from 'vscode';
+import { Telemetry } from '../../datascience/constants';
+import { EnvironmentType, PythonEnvironment } from '../../pythonEnvironments/info';
+import { sendTelemetryEvent } from '../../telemetry';
+import { IApplicationShell, IWorkspaceService } from '../application/types';
+import { wrapCancellationTokens } from '../cancellation';
+import { STANDARD_OUTPUT_CHANNEL } from '../constants';
+import { disposeAllDisposables } from '../helpers';
+import { traceError, traceVerbose, traceWarning } from '../logger';
+import { getDisplayPath } from '../platform/fs-paths';
+import { IPythonExecutionFactory } from '../process/types';
+import { IDisposable, IOutputChannel, Product, Resource } from '../types';
+import { DataScience } from '../utils/localize';
+import { translateProductToModule } from './productInstaller';
+import { ProductNames } from './productNames';
+
+/**
+ * Class used to install IPyKernel into global enviroments if Python extension fails to install it for what ever reason.
+ * We know ipykernel can be easily installed with `python -m pip install ipykernel` for Global Python Environments.
+ * Similarly for Jupyter & other python packages.
+ *
+ * Note: This is only a fallback, we know that sometimes Python fails to install these & Python installs them via the terminal.
+ * We don't really know why it fails to install these packages.
+ */
+@injectable()
+export class BackupPipInstaller {
+    constructor(
+        @inject(IApplicationShell) private readonly appShell: IApplicationShell,
+        @inject(IWorkspaceService) private readonly workspaceService: IWorkspaceService,
+        @inject(IOutputChannel) @named(STANDARD_OUTPUT_CHANNEL) private readonly outputChannel: OutputChannel,
+        @inject(IPythonExecutionFactory) private readonly pythonExecFactory: IPythonExecutionFactory,
+        private readonly isInstalled: (product: Product, interpreter: PythonEnvironment) => Promise<boolean | undefined>
+    ) {}
+    public async install(
+        product: Product,
+        interpreter: PythonEnvironment,
+        resource: Resource,
+        reInstallAndUpdate: boolean,
+        token: CancellationToken
+    ): Promise<boolean> {
+        if (product === Product.pip) {
+            return false;
+        }
+        // We can only run this installer against global enviorments.
+        if (interpreter.envType !== EnvironmentType.Global) {
+            traceWarning(`We cannot pip install packages into non-Global Python environments.`);
+            return false;
+        }
+        // Check if pip is installed.
+        const isPipInstalled = await this.isInstalled(Product.pip, interpreter);
```

Comment:
For now we're only going to install DS pacakges via Pip for Global & virtual python environments.